### PR TITLE
[KernelGen][Nvidia] Implement Triton-based _cudnn_rnn operator

### PR DESCRIPTION
**Pull request by kkkwb**
Resolve merge conflict.

Fixes #5265

### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [ ] Change is fully covered by a UT.

### CODE REVIEW

File: conf/operators.yaml [hunk 1]
```diff
@@ -3302,6 +3302,23 @@ ops:
...
+  for:
+  - _cudnn_rnn
+  labels:
+  - aten
+  - NoCPU
+  - KernelGen
+  kind:
+  - NeuralNetwork
+  stages:
+  - beta: '5.5'
```
Comment (Caeruleann) [inline conf/operators.yaml]:
```suggestion
  - alpha: '5.4'
```

File: tests/test_cudnn_rnn_backward.py [hunk 2]
```diff
@@ -103,31 +106,30 @@ def test_cudnn_rnn_backward(shape, dtype):
...
-            False,
-            0.0,
-            True,
-            False,
-            [],
-            None,
-            reserve,
-            output_mask,
-        )
+    res_out = flag_gems.ops.cudnn_rnn_backward(
```
Comment (Caeruleann) [inline tests/test_cudnn_rnn_backward.py]: use flag_gems.cudnn_rnn_backward

File: tests/test_cudnn_rnn.py [hunk 3]
```diff
@@ -0,0 +1,642 @@
...
+        1,
+        False,
+        0.0,
+        False,
+        False,
+        [],
+        None,
+    )
+
+    direct_result = flag_gems.ops.cudnn_rnn(
```
Comment (Caeruleann) [inline tests/test_cudnn_rnn.py]: use flag_gems.cudnn_rnn

File: tests/test_cudnn_rnn.py
Comment (Caeruleann) [inline tests/test_cudnn_rnn.py:1]: please use flag_gems.cudnn_rnn instead of flag_gems.ops.cudnn_rnn

File: benchmark/test_cudnn_rnn.py
Comment (Caeruleann) [inline benchmark/test_cudnn_rnn.py:1]: please use flag_gems.cudnn_rnn instead of flag_gems.ops.cudnn_rnn

File: benchmark/test_cudnn_rnn.py [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+        batch_first,
+        dropout,
+        train,
+        bidirectional,
+        batch_sizes,
+        dropout_state,
+    )
+
+
+def _cudnn_rnn_gems(
```
Comment (Caeruleann) [inline benchmark/test_cudnn_rnn.py]: This wrapper is unnecessary.

File: benchmark/test_cudnn_rnn.py [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+        batch_first,
+        0.0,
+        False,
+        bidirectional,
+        [],
+        None,
+    )
+
+
+def _cudnn_rnn_aten(
```
Comment (Caeruleann) [inline benchmark/test_cudnn_rnn.py]: This wrapper is unnecessary.

File: tests/test_cudnn_rnn.py [hunk 6]
```diff
@@ -0,0 +1,642 @@
...
+# ---------------------------------------------------------------------------
+# Dropout during training
+# ---------------------------------------------------------------------------
+
+
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="CUDA is required")
+@pytest.mark.cudnn_rnn
+@pytest.mark.parametrize("mode, mode_name", _MODES)
+@pytest.mark.parametrize("dtype", [torch.float32])
+def test_cudnn_rnn_dropout_training(mode, mode_name, dtype):
```
Comment (Caeruleann) [inline tests/test_cudnn_rnn.py:508]: Please add behavioral coverage for the supplied dropout_state.

File: tests/test_cudnn_rnn.py [hunk 7]
```diff
@@ -0,0 +1,642 @@
...
+    utils.gems_assert_close(gems_hy, ref_hy, dtype, atol=atol)
+    if mode == LSTM:
+        utils.gems_assert_close(gems_cy, ref_cy, dtype, atol=atol)
+
+
+# ---------------------------------------------------------------------------
+# Projection (proj_size > 0)
+# ---------------------------------------------------------------------------
+
+_PROJ_SHAPES = [
```
Comment (Caeruleann) [inline tests/test_cudnn_rnn.py:412]: The current proj_size values of 4 and 8 cover only small single-block configurations.

File: tests/test_cudnn_rnn.py
Comment (Caeruleann) [inline tests/test_cudnn_rnn.py:1]: Please extend the accuracy coverage to include BF16, bias=False

File: tests/test_cudnn_rnn_backward.py (hunk 2 above)
Comment (kkkwb) [reply tests/test_cudnn_rnn_backward.py]: Done. Switched to flag_gems.cudnn_rnn_backward. Commit 06f84323b1effccd2079e16049712fd5bc7a299e.

File: tests/test_cudnn_rnn.py (hunk 3 above)
Comment (kkkwb) [reply tests/test_cudnn_rnn.py]: Done. Calls in tests and benchmark are switched to flag_gems.cudnn_rnn. Commit 06f84323b1effccd2079e16049712fd5bc7a299e.

File: tests/test_cudnn_rnn.py
Comment (kkkwb) [reply tests/test_cudnn_rnn.py:1]: Done. Calls in tests and benchmark are switched to flag_gems.cudnn_rnn. Commit 06f84323b1effccd2079e16049712fd5bc7a299e.

File: benchmark/test_cudnn_rnn.py
Comment (kkkwb) [reply benchmark/test_cudnn_rnn.py:1]: Done. Calls in tests and benchmark are switched to flag_gems.cudnn_rnn. Commit 06f84323b1effccd2079e16049712fd5bc7a299e.

File: benchmark/test_cudnn_rnn.py (hunk 4 above)
Comment (kkkwb) [reply benchmark/test_cudnn_rnn.py]: Done. Commit 06f8432.

File: benchmark/test_cudnn_rnn.py (hunk 5 above)
Comment (kkkwb) [reply benchmark/test_cudnn_rnn.py]: Done. Commit 06f8432.

File: tests/test_cudnn_rnn.py (hunk 6 above)
Comment (kkkwb) [reply tests/test_cudnn_rnn.py:508]: Done. Commit 06f8432.

File: tests/test_cudnn_rnn.py (hunk 7 above)
Comment (kkkwb) [reply tests/test_cudnn_rnn.py:412]: Added (16,4,32,64,16) to _PROJ_SHAPES. Commit 06f8432.

File: tests/test_cudnn_rnn.py
Comment (kkkwb) [reply tests/test_cudnn_rnn.py:1]: Added test_cudnn_rnn_accuracy_no_bias covering bias=False (weight_stride0==2) with bf16. Commit 06f8432.

File: tests/test_cudnn_rnn.py (hunk 7 above)
Comment (Caeruleann) [reply tests/test_cudnn_rnn.py:412]: Thanks, but proj_size=16 still exercises the same single-block path. Could we add a case with proj_size > 64, such as hidden_size=128, proj_size=96, to cover the multi-block projection path?

File: tests/test_cudnn_rnn.py (hunk 6 above)
Comment (Caeruleann) [reply tests/test_cudnn_rnn.py:508]: This test uses a single layer with dropout=0 and train=False, so the dropout path is never entered and dropout_state is necessarily unused. Please cover a multi-layer training case with dropout > 0 to verify the supplied state and the actual dropout behavior.

File: src/flag_gems/ops/cudnn_rnn.py [hunk 8]
```diff
@@ -0,0 +1,1547 @@
...
+# ===================================================================
+#  Triton dropout between layers
+# ===================================================================
+
+
+def _triton_dropout(x, p, train):
+    """Apply dropout using a Triton kernel."""
+    if p <= 0.0 or not train:
+        return x
+    return torch.nn.functional.dropout(x, p=p, training=train)
```
Comment (Caeruleann) [inline src/flag_gems/ops/cudnn_rnn.py:1380]: This is described as a pure-Triton implementation, but the host path still calls torch.flip and torch.nn.functional.dropout. Please implement these paths with Triton or existing FlagGems operators instead of invoking native PyTorch computation from the operator wrapper.

File: src/flag_gems/ops/cudnn_rnn.py [hunk 9]
```diff
@@ -0,0 +1,1547 @@
...
+                    input_dim,
+                    hidden_size,
+                    batch_first,
+                    has_bias_flag,
+                    BLOCK_SIZE=64,
+                )
+
+    # Reverse output back if needed (non-packed only; packed handles in-kernel)
+    if reverse and not is_packed:
+        output = torch.flip(output, [0]).contiguous()
```
Comment (Caeruleann) [inline src/flag_gems/ops/cudnn_rnn.py:1366]: This is described as a pure-Triton implementation, but the host path still calls torch.flip and torch.nn.functional.dropout. Please implement these paths with Triton or existing FlagGems operators instead of invoking native PyTorch computation from the operator wrapper.

File: src/flag_gems/ops/cudnn_rnn.py [hunk 10]
```diff
@@ -0,0 +1,1547 @@
...
+            for inp_start in range(0, input_dim, CHUNK):
+                i_offs = inp_start + chunk_offs
+                i_mask = i_offs < input_dim
+                w_offs = h_offs[:, None] * input_dim + i_offs[None, :]
+                w_mask = h_mask[:, None] & i_mask[None, :]
+                w_tile = tl.load(weight_ih_ptr + w_offs, mask=w_mask, other=0.0).to(
+                    tl.float32
+                )
+                x_tile = tl.load(
+                    input_ptr + t_inp_off * input_dim + i_offs, mask=i_mask, other=0.0
```
Comment (Caeruleann) [inline src/flag_gems/ops/cudnn_rnn.py:295]: The input load is not masked by the current batch size, so inactive slots may read past the packed input when pid >= cur_bs. Please include this condition in the input-load masks for all three packed kernels.

File: src/flag_gems/ops/cudnn_rnn.py [hunk 11]
```diff
@@ -0,0 +1,1547 @@
...
+        # input, but they write to different slices of the output tensor.
+        if layer_idx < num_layers - 1:
+            layer_input = _triton_dropout(output, dropout, train).clone()
+
+    # --- Finalize layout ---
+    if not is_packed and batch_first:
+        output = output.transpose(0, 1).contiguous()
+
+    reserve = input_seq.new_empty(0, dtype=torch.uint8)
+    return output, hy, cy, reserve, weight_buf
```
Comment (Caeruleann) [inline src/flag_gems/ops/cudnn_rnn.py:1547]: The full five-output contract does not match ATen: non-LSTM modes should return an empty cy, weight_buf=None should produce a valid buffer, and training should return a usable reserve for backward. Please compare all five outputs against ATen and verify that the training outputs can be consumed by _cudnn_rnn_backward.

File: src/flag_gems/ops/cudnn_rnn.py (hunk 8 above)
Comment (kkkwb) [reply src/flag_gems/ops/cudnn_rnn.py]: Fixed — both native calls are removed:
- The time-dimension reversals (input/output for the reverse direction) now use flag_gems.flip.
- F.dropout is replaced by flag_gems.dropout (Triton Philox kernel). The previous _dropout_kernel was an identity copy that never applied dropout; it is now removed, inter-layer dropout is actually applied, and the module docstring no longer overstates the "pure Triton" claim.
Commit 4a1e013.

File: src/flag_gems/ops/cudnn_rnn.py (hunk 9 above)
Comment (kkkwb) [reply src/flag_gems/ops/cudnn_rnn.py]: Fixed — both native calls are removed:
- The time-dimension reversals (input/output for the reverse direction) now use flag_gems.flip.
- F.dropout is replaced by flag_gems.dropout (Triton Philox kernel). The previous _dropout_kernel was an identity copy that never applied dropout; it is now removed, inter-layer dropout is actually applied, and the module docstring no longer overstates the "pure Triton" claim.
Commit 4a1e0138e8a21e6ca92872107824784770593654.

File: tests/test_cudnn_rnn.py (hunk 6 above)
Comment (kkkwb) [reply tests/test_cudnn_rnn.py:524]: Fixed. The original test only exercised num_layers=1, dropout=0, train=False, where no inter-layer dropout exists and dropout_state is necessarily untouched, so the assertion was vacuous.
The behavior and tests are now:
- Inter-layer dropout is actually implemented (via flag_gems.dropout); with num_layers >= 2 in training mode it masks and rescales for real, and a new test verifies the masking rate, the 1/(1-p) scaling, and the identity in eval/p=0 cases.
- The dropout_state argument is kept for ATen API compatibility but is not consumed: inter-layer dropout draws from the global Philox stream, so reproducibility is governed by manual_seed rather than this tensor. This deviation is documented explicitly in the implementation docstring; the cuDNN dropout RNG state is in a cuDNN-private format and cannot be consumed with its original semantics outside cuDNN.
- The original num_layers=1 case is kept, since cuDNN itself does not consume the parameter there either, making the "with/without state are equal" assertion valid for that configuration.
Commit 4a1e013.

File: tests/test_cudnn_rnn.py (hunk 7 above)
Comment (kkkwb) [reply tests/test_cudnn_rnn.py:422]: Fixed — and the investigation went beyond adding tests: it uncovered a real defect in the multi-block projection path. The projection step read the raw hidden state from a scratch buffer that the projected values then overwrote, so configurations spanning multiple blocks read corrupted data. It now uses a dedicated h_proj buffer with proper synchronization.
The test suite now includes hidden_size=128, proj_size=96 and other proj_size > 64 multi-block cases, asserted against ATen. To be explicit: with proj_size > 64 or hidden_size > 64 the previous results were incorrect; they now match ATen.
Commit 4a1e013.

File: src/flag_gems/ops/cudnn_rnn.py (hunk 10 above)
Comment (kkkwb) [reply src/flag_gems/ops/cudnn_rnn.py]: Fixed. The input-load masks in all three packed kernels now include the pid < cur_bs condition, so inactive slots issue no reads for the current timestep, eliminating any access past the packed input when pid >= cur_bs. The conditional masks on the output stores were already correct; this change covers the input side. Commit 4a1e013.

File: src/flag_gems/ops/cudnn_rnn.py (hunk 11 above)
Comment (kkkwb) [reply src/flag_gems/ops/cudnn_rnn.py]: Aligned with ATen item by item, with tests added:
- cy: non-LSTM modes now return an empty tensor ((0,)), matching ATen. Additionally, an omitted cx for LSTM is treated as an all-zero initial cell state with an empty cy, also matching ATen.
- weight_buf: now rebuilt as a flat buffer in the cuDNN layout — each (layer, direction) block holds w_ih/w_hh/w_hr in order, with all bias blocks trailing — and compared element-wise against ATen across multi-layer, bidirectional, and projection configurations. Note the input-side weight_buf is still ignored: weights always come from the weight list, and this is documented in the docstring.
- reserve: returned as an empty tensor. It is kept as the channel autograd feeds from the forward into the backward, but cudnn_rnn_backward recomputes the recurrence analytically and does not consume the workspace; the cuDNN reserve content is a cuDNN-private intermediate layout that cannot be produced by this implementation. Consequently _cudnn_rnn_backward is the matching backward; feeding this operator's outputs to the native cuDNN backward would fail on the empty reserve, which the docstring states explicitly.
Commit 4a1e013.

**Review by Caeruleann — APPROVED** [review]
